### PR TITLE
feat: add support for remapping ivf_pq indices

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
Note: the current approach uses tombstones to keep the index the exact same size as it was before.  This was done for two reasons:

 * No need to modify the index offsets
 * It's difficult to know exactly where the metadata starts in an IVF file
 
However, using tombstones means we have to do slightly more work in the PQ search (to remove the invalid rows).  There are some other approaches we could take:

 * We could go ahead and shrink the pages (would need to update offsets).  To deal with the fact that we don't know the exact metadata boundary we could still copy the entire file.  This means we'd have some garbage between the pages and the metadata.
 * We could start storing an offset to the metadata.  This would mean a change to the index format (old indices would be incompatible).  Then we should be able to shrink the pages without creating garbage.
 
 
Untitled-2023-09-22-0819

File: rust/lance/src/index/vector/pq.rs [hunk 1]
```diff
@@ -249,6 +250,60 @@ impl Index for PQIndex {
     }
 }
 
+// Helper struct for zipped distance + row id that sorts by distance
+struct DistanceRowId((f32, u64));
+
+impl DistanceRowId {
+    fn distance(&self) -> f32 {
```
Comment (eddyxu) [inline rust/lance/src/index/vector/pq.rs:257]: Shall we just use

```
struct DistanceRowId {
pub  distance: f32
pub row_id: u64
}
```
For readability?

File: rust/lance/src/index/vector/pq.rs [hunk 2]
```diff
@@ -277,10 +332,28 @@ impl VectorIndex for PQIndex {
...
+        // Remove any tombstone rows from consideration when sorting
+        let distance_ids = distances
+            .as_any()
+            .downcast_ref::<Float32Array>()
+            .unwrap()
+            .values()
+            .iter()
+            .copied()
+            .zip(row_ids.values().iter().copied())
+            .filter(|(_, row_id)| *row_id != RowId::TOMBSTONE_ROW)
```
Comment (eddyxu) [inline rust/lance/src/index/vector/pq.rs:344]: Is this before remapping is done or after?

If the remapping is done, why there still are tomestone rows in the index file?

File: rust/lance/src/index/vector/pq.rs [hunk 3]
```diff
@@ -318,6 +391,29 @@ impl VectorIndex for PQIndex {
...
+    fn remap(&mut self, mapping: &HashMap<u64, Option<u64>>) -> Result<()> {
+        let remapped_ids =
+            UInt64Array::from_iter_values(self.row_ids.as_ref().unwrap().values().iter().map(
+                |old_row_id| {
+                    mapping
+                        .get(old_row_id)
+                        .cloned()
+                        // If the row id is not in the mapping then this row is not remapped and we keep as is
+                        .unwrap_or(Some(*old_row_id))
+                        // If the row is in the mapping, but maps to None, then it is deleted, and we insert
```
Comment (eddyxu) [inline rust/lance/src/index/vector/pq.rs:408]: If the row is deleted , let's just not throw it away?  IIUC, putting a tombstone in index does not help any process later to clean up this row?

File: rust/lance/src/index.rs [hunk 4]
```diff
@@ -101,6 +103,12 @@ pub trait DatasetIndexExt {
         params: &dyn IndexParams,
         replace: bool,
     ) -> Result<()>;
+
+    async fn remap_index(
```
Comment (eddyxu) [inline rust/lance/src/index.rs:107]: Let's just make a `optimize_index()` method to contain this remapping, deletion, or later incremental update together?

From the user's aspect, it is hard for them to understand what impact of exact each action is.  An general `optimize_index()` for them would be "call this method and your index will be faster"

Comment (westonpace) [reply rust/lance/src/index.rs:107]: The user would never call this method (they would not be able to provide `row_id_map`).  I see your point though. I will move this out of `DatasetIndexExt` and call straight into the lower level method later.

File: rust/lance/src/index/vector.rs [hunk 5]
```diff
@@ -268,6 +269,27 @@ pub(crate) async fn build_vector_index(
...
+pub(crate) async fn remap_vector_index(
+    dataset: Arc<Dataset>,
+    column: &str,
+    old_uuid: &Uuid,
+    new_uuid: &Uuid,
+    mapping: &HashMap<u64, Option<u64>>,
+) -> Result<()> {
+    let old_index = open_index(dataset.clone(), column, &old_uuid.to_string()).await?;
+    old_index.check_can_remap()?;
+    let ivf_index: &IVFIndex = old_index.as_any().downcast_ref().unwrap();
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/index/vector.rs]: I now return an error here instead:

```
    let ivf_index: &IVFIndex =
        old_index
            .as_any()
            .downcast_ref()
            .ok_or_else(|| Error::NotSupported {
                source: "Only IVF indexes can be remapped currently".into(),
            })?;
```

File: rust/lance/src/index/vector/pq.rs (hunk 1 above)
Comment (westonpace) [reply rust/lance/src/index/vector/pq.rs:266]: Done

**Comment by westonpace** [comment]
I made an attempt this morning to remove the tombstone rows and I would like to defer this for a bit.  It involved making a bunch of assertions as I read the metadata back in (must be two stages, first ivf, then pq) and I want to instead refactor some of the approach to be more general.

I'm going to merge this into the feature branch, then get the compaction PR in place, and then try and tackle index remapping in a more generic way (so it walks through and remaps each of the stages, updating the metadata as it goes) so I can remove the tombstone code.

**Pull request merged**